File: Cpp-C/Ema/Src/Access/Impl/ServiceEndpointDiscoveryImpl.cpp
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license      --
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
 *|                See the project's LICENSE.md for details.                  --
 *|           Copyright (C) 2019 Refinitiv. All rights reserved.            --
 *|-----------------------------------------------------------------------------
 */

#include "ServiceEndpointDiscoveryImpl.h"
#include "ExceptionTranslator.h"
#include "Utilities.h"
#include "OmmInvalidUsageException.h"

using namespace rtsdk::ema::access;

ServiceEndpointDiscoveryImpl::ServiceEndpointDiscoveryImpl(ServiceEndpointDiscovery *pServiceEndpointDiscovery, const EmaString& tokenServiceURL, 
																const EmaString& serviceDiscoveryURL) :
_pServiceEndpointDiscovery(0),
_pClient(0)
{
	RsslCreateReactorOptions reactorOpts;
	RsslErrorInfo rsslErrorInfo;
	clearRsslErrorInfo(&rsslErrorInfo);

	rsslClearCreateReactorOptions(&reactorOpts);
	reactorOpts.tokenServiceURL.data = const_cast<char*>(tokenServiceURL.c_str());
	reactorOpts.tokenServiceURL.length = tokenServiceURL.length();
	reactorOpts.serviceDiscoveryURL.data = const_cast<char*>(serviceDiscoveryURL.c_str());
	reactorOpts.serviceDiscoveryURL.length = serviceDiscoveryURL.length();
	reactorOpts.userSpecPtr = this;

	_pReactor = rsslCreateReactor(&reactorOpts, &rsslErrorInfo);
	if (!_pReactor)
	{
		EmaString temp("Failed to initialize ServiceEndpointDiscoveryImpl (rsslCreateReactor).");
		temp.append("' Error Id='").append(rsslErrorInfo.rsslError.rsslErrorId)
			.append("' Internal sysError='").append(rsslErrorInfo.rsslError.sysError)
			.append("' Error Location='").append(rsslErrorInfo.errorLocation)
			.append("' Error Text='").append(rsslErrorInfo.rsslError.text).append("'. ");
		throwIueException( temp, OmmInvalidUsageException::InternalErrorEnum );
	}

	_pServiceEndpointDiscovery = pServiceEndpointDiscovery;
}

ServiceEndpointDiscoveryImpl::~ServiceEndpointDiscoveryImpl()
{
	if (_pReactor)
	{
		RsslErrorInfo rsslErrorInfo;
		clearRsslErrorInfo(&rsslErrorInfo);
		rsslDestroyReactor(_pReactor, &rsslErrorInfo);

		_pReactor = 0;
	}
}

void ServiceEndpointDiscoveryImpl::registerClient(const ServiceEndpointDiscoveryOption& params, ServiceEndpointDiscoveryClient& client, 
														void *closure)
{
	_userLock.lock();

	RsslErrorInfo rsslErrorInfo;
	clearRsslErrorInfo(&rsslErrorInfo);
	rsslClearReactorServiceDiscoveryOptions(&_serviceDiscoveryOpts);

	_serviceDiscoveryOpts.userName.data = const_cast<char*>(params._username.c_str());
	_serviceDiscoveryOpts.userName.length = params._username.length();
	_serviceDiscoveryOpts.password.data = const_cast<char*>(params._password.c_str());
	_serviceDiscoveryOpts.password.length = params._password.length();
	_serviceDiscoveryOpts.takeExclusiveSignOnControl = (RsslBool)params._takeExclusiveSignOnControl;

	if (!params._proxyHostName.empty())
	{
		_serviceDiscoveryOpts.proxyHostName.data = const_cast<char*>(params._proxyHostName.c_str());
		_serviceDiscoveryOpts.proxyHostName.length = params._proxyHostName.length();
	}

	if (!params._proxyPort.empty())
	{
		_serviceDiscoveryOpts.proxyPort.data = const_cast<char*>(params._proxyPort.c_str());
		_serviceDiscoveryOpts.proxyPort.length = params._proxyPort.length();
	}

	if (!params._proxyUserName.empty())
	{
		_serviceDiscoveryOpts.proxyUserName.data = const_cast<char*>(params._proxyUserName.c_str());
		_serviceDiscoveryOpts.proxyUserName.length = params._proxyUserName.length();
	}

	if (!params._proxyPassword.empty())
	{
		_serviceDiscoveryOpts.proxyPasswd.data = const_cast<char*>(params._proxyPassword.c_str());
		_serviceDiscoveryOpts.proxyPasswd.length = params._proxyPassword.length();
	}

	if (!params._proxyDomain.empty())
	{
		_serviceDiscoveryOpts.proxyDomain.data = const_cast<char*>(params._proxyDomain.c_str());
		_serviceDiscoveryOpts.proxyDomain.length = params._proxyDomain.length();
	}

	if (!params._clientId.empty())
	{
		_serviceDiscoveryOpts.clientId.data = const_cast<char*>(params._clientId.c_str());
		_serviceDiscoveryOpts.clientId.length = params._clientId.length();
	}

	if (!params._clientSecret.empty())
	{
		_serviceDiscoveryOpts.clientSecret.data = const_cast<char*>(params._clientSecret.c_str());
		_serviceDiscoveryOpts.clientSecret.length = params._clientSecret.length();
	}

	if (!params._tokenScope.empty() && params._tokenScope != _serviceDiscoveryOpts.tokenScope.data)
	{
		_serviceDiscoveryOpts.tokenScope.data = const_cast<char*>(params._tokenScope.c_str());
		_serviceDiscoveryOpts.tokenScope.length = params._tokenScope.length();
	}

	_pClient = &client;
	_serviceDiscoveryOpts.userSpecPtr = closure;
	_serviceDiscoveryOpts.pServiceEndpointEventCallback = ServiceEndpointDiscoveryImpl::serviceEndpointEventCallback;

	EmaString temp;
	switch (params._transport)
	{
	case ServiceEndpointDiscoveryOption::UnknownTransportEnum:
		break;
	case ServiceEndpointDiscoveryOption::TcpEnum:
		_serviceDiscoveryOpts.transport = RSSL_RD_TP_TCP;
		break;
	case ServiceEndpointDiscoveryOption::WebsocketEnum:
		_serviceDiscoveryOpts.transport = RSSL_RD_TP_WEBSOCKET;
		break;
	default:
		temp.set("Invalid transport protocol ").append(params._transport)
			.append(" specified in ServiceEndpointDiscoveryOption::transport()");

		_userLock.unlock();
		throwIueException( temp, OmmInvalidUsageException::InvalidArgumentEnum );
		break;
	}

	switch (params._dataFormat)
	{
	case ServiceEndpointDiscoveryOption::UnknownDataFormatEnum:
		break;
	case ServiceEndpointDiscoveryOption::RwfEnum:
		_serviceDiscoveryOpts.dataFormat = RSSL_RD_DP_RWF;
		break;
	case ServiceEndpointDiscoveryOption::Json2Enum:
		_serviceDiscoveryOpts.dataFormat = RSSL_RD_DP_JSON2;
		break;
	default:
		temp.set("Invalid dataformat protocol ").append(params._dataFormat)
			.append(" specified in ServiceEndpointDiscoveryOption::dataFormat()");

		_userLock.unlock();
		throwIueException( temp, OmmInvalidUsageException::InvalidArgumentEnum );
		break;
	}

	if (rsslReactorQueryServiceDiscovery(_pReactor, &_serviceDiscoveryOpts, &rsslErrorInfo) != RSSL_RET_SUCCESS)
	{
		EmaString temp("Failed to query service discovery (rsslReactorQueryServiceDiscovery).");
		temp.append("' Error Id='").append(rsslErrorInfo.rsslError.rsslErrorId)
			.append("' Internal sysError='").append(rsslErrorInfo.rsslError.sysError)
			.append("' Error Location='").append(rsslErrorInfo.errorLocation)
			.append("' Error Text='").append(rsslErrorInfo.rsslError.text).append("'. ");

		_userLock.unlock();
		throwIueException( temp, rsslErrorInfo.rsslError.rsslErrorId );
	}

	_userLock.unlock();
}

RsslReactorCallbackRet ServiceEndpointDiscoveryImpl::serviceEndpointEventCallback(RsslReactor* pReactor, RsslReactorServiceEndpointEvent* pEvent)
{
	UInt32 index, valueIndex;
	ServiceEndpointDiscoveryImpl *pImpl = (ServiceEndpointDiscoveryImpl*)pReactor->userSpecPtr;
	RsslReactorServiceEndpointInfo *reactorServiceEndpointInfo;
	ServiceEndpointDiscoveryInfo serviceEndpointInfo;
	EmaString value;

	pImpl->_serviceDiscoveryEvent._pClosure = pEvent->userSpecPtr;
	pImpl->_serviceDiscoveryEvent._pServiceEndpointDiscovery = pImpl->_pServiceEndpointDiscovery;

	if (pEvent->pErrorInfo == 0)
	{
		pImpl->_serviceDiscoveryResp._pServiceEndpointDiscoveryInfoList->clear();
		for (index = 0; index < pEvent->serviceEndpointInfoCount; index++)
		{
			reactorServiceEndpointInfo = &pEvent->serviceEndpointInfoList[index];

			serviceEndpointInfo._pDataFormatList->clear();
			serviceEndpointInfo._pLocationList->clear();
			for (valueIndex = 0; valueIndex < reactorServiceEndpointInfo->dataFormatCount; valueIndex++)
			{
				value.set(reactorServiceEndpointInfo->dataFormatList[valueIndex].data, reactorServiceEndpointInfo->dataFormatList[valueIndex].length);
				serviceEndpointInfo._pDataFormatList->push_back(value);
			}

			for (valueIndex = 0; valueIndex < reactorServiceEndpointInfo->locationCount; valueIndex++)
			{
				value.set(reactorServiceEndpointInfo->locationList[valueIndex].data, reactorServiceEndpointInfo->locationList[valueIndex].length);
				serviceEndpointInfo._pLocationList->push_back(value);
			}

			serviceEndpointInfo._endPoint.set(reactorServiceEndpointInfo->endPoint.data, reactorServiceEndpointInfo->endPoint.length);
			serviceEndpointInfo._port.set(reactorServiceEndpointInfo->port.data, reactorServiceEndpointInfo->port.length);
			serviceEndpointInfo._provider.set(reactorServiceEndpointInfo->provider.data, reactorServiceEndpointInfo->provider.length);
			serviceEndpointInfo._transport.set(reactorServiceEndpointInfo->transport.data, reactorServiceEndpointInfo->provider.length);
			pImpl->_serviceDiscoveryResp._pServiceEndpointDiscoveryInfoList->push_back(serviceEndpointInfo);
		}

		static_cast<ServiceEndpointDiscoveryImpl*>(pReactor->userSpecPtr)->_pClient->onSuccess(pImpl->_serviceDiscoveryResp, pImpl->_serviceDiscoveryEvent);
	}
	else
	{
		static_cast<ServiceEndpointDiscoveryImpl*>(pReactor->userSpecPtr)->_pClient->onError(value.set(pEvent->pErrorInfo->rsslError.text), pImpl->_serviceDiscoveryEvent);
	}

	return RSSL_RC_CRET_SUCCESS;
}

File: Java/Ema/Core/src/main/java/com/thomsonreuters/ema/access/ServiceEndpointDiscoveryImpl.java
///*|-----------------------------------------------------------------------------
// *|            This source code is provided under the Apache 2.0 license      --
// *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
// *|                See the project's LICENSE.md for details.                  --
// *|           Copyright (C) 2019 Refinitiv. All rights reserved.            --
///*|-----------------------------------------------------------------------------

package com.thomsonreuters.ema.access;
import java.util.concurrent.locks.ReentrantLock;

import com.thomsonreuters.ema.access.ServiceEndpointDiscovery;
import com.thomsonreuters.ema.access.ServiceEndpointDiscoveryClient;
import com.thomsonreuters.ema.access.ServiceEndpointDiscoveryOption;
import com.thomsonreuters.upa.codec.Buffer;
import com.thomsonreuters.upa.codec.CodecFactory;
import com.thomsonreuters.upa.valueadd.reactor.Reactor;
import com.thomsonreuters.upa.valueadd.reactor.ReactorCallbackReturnCodes;
import com.thomsonreuters.upa.valueadd.reactor.ReactorDiscoveryDataFormatProtocol;
import com.thomsonreuters.upa.valueadd.reactor.ReactorDiscoveryTransportProtocol;
import com.thomsonreuters.upa.valueadd.reactor.ReactorErrorInfo;
import com.thomsonreuters.upa.valueadd.reactor.ReactorFactory;
import com.thomsonreuters.upa.valueadd.reactor.ReactorOptions;
import com.thomsonreuters.upa.valueadd.reactor.ReactorReturnCodes;
import com.thomsonreuters.upa.valueadd.reactor.ReactorServiceDiscoveryOptions;
import com.thomsonreuters.upa.valueadd.reactor.ReactorServiceEndpointEvent;
import com.thomsonreuters.upa.valueadd.reactor.ReactorServiceEndpointEventCallback;
import com.thomsonreuters.upa.valueadd.reactor.ReactorServiceEndpointInfo;

class ServiceEndpointDiscoveryImpl implements ServiceEndpointDiscovery, ReactorServiceEndpointEventCallback
{
	private Reactor _reactor;
	private ReactorErrorInfo _reactorErrorInfo = ReactorFactory.createReactorErrorInfo();
	private StringBuilder _strBuilder;
	private ReentrantLock _userLock = new java.util.concurrent.locks.ReentrantLock();
	private ReactorServiceDiscoveryOptions _reactorServiceDiscoveryOptions = ReactorFactory.createReactorServiceDiscoveryOptions();
	private OmmInvalidUsageExceptionImpl _ommIUExcept;
	private ServiceEndpointDiscoveryRespImpl _serviceEndpointDiscoveryRespImpl = new ServiceEndpointDiscoveryRespImpl();
	private ServiceEndpointDiscoveryEventImpl _ServiceEndpointDiscoveryEventImpl = new ServiceEndpointDiscoveryEventImpl();
	private ServiceEndpointDiscoveryClient _client;
	private Buffer _tokenServiceURLBuf = CodecFactory.createBuffer();
	private Buffer _serviceDiscoveryUrlBuf = CodecFactory.createBuffer();
	
	ServiceEndpointDiscoveryImpl()
	{
		ReactorOptions reactorOptions = ReactorFactory.createReactorOptions();
		reactorOptions.clear();
		_reactorErrorInfo.clear();
		
		reactorOptions.userSpecObj(this);
		_reactor = ReactorFactory.createReactor(reactorOptions, _reactorErrorInfo);
		if(_reactorErrorInfo.code() != ReactorReturnCodes.SUCCESS)
		{
			strBuilder().append("Failed to create ServiceEndpointDiscoveryImpl (ReactorFactory.createReactor).")
			.append("' Error Id='").append(_reactorErrorInfo.error().errorId()).append("' Internal sysError='")
			.append(_reactorErrorInfo.error().sysError()).append("' Error Location='")
			.append(_reactorErrorInfo.location()).append("' Error Text='")
			.append(_reactorErrorInfo.error().text()).append("'. ");
			
			throw ommIUExcept().message(_strBuilder.toString(), OmmInvalidUsageException.ErrorCode.INTERNAL_ERROR);
		}
	}
	
	ServiceEndpointDiscoveryImpl(String tokenServiceUrl, String serviceDiscoveryUrl)
	{
		ReactorOptions reactorOptions = ReactorFactory.createReactorOptions();
		reactorOptions.clear();
		_reactorErrorInfo.clear();
		
		reactorOptions.userSpecObj(this);
		
		if ((tokenServiceUrl != null) && (!tokenServiceUrl.isEmpty()))
		{
			_tokenServiceURLBuf.data(tokenServiceUrl);
			reactorOptions.tokenServiceURL(_tokenServiceURLBuf);
		}
		
		if((serviceDiscoveryUrl != null) && (!serviceDiscoveryUrl.isEmpty()))
		{
			_serviceDiscoveryUrlBuf.data(serviceDiscoveryUrl);
			reactorOptions.serviceDiscoveryURL(_serviceDiscoveryUrlBuf);
		}
		
		_reactor = ReactorFactory.createReactor(reactorOptions, _reactorErrorInfo);
		if(_reactorErrorInfo.code() != ReactorReturnCodes.SUCCESS)
		{
			strBuilder().append("Failed to create ServiceEndpointDiscoveryImpl (ReactorFactory.createReactor).")
			.append("' Error Id='").append(_reactorErrorInfo.error().errorId()).append("' Internal sysError='")
			.append(_reactorErrorInfo.error().sysError()).append("' Error Location='")
			.append(_reactorErrorInfo.location()).append("' Error Text='")
			.append(_reactorErrorInfo.error().text()).append("'. ");
			
			throw ommIUExcept().message(_strBuilder.toString(), OmmInvalidUsageException.ErrorCode.INTERNAL_ERROR);
		}
	}
	
	@Override
	public void registerClient(ServiceEndpointDiscoveryOption params, ServiceEndpointDiscoveryClient client)
	{
		registerClient(params, client, null);
	}

	@Override
	public void registerClient(ServiceEndpointDiscoveryOption params, ServiceEndpointDiscoveryClient client, Object closure)
	{
		_userLock.lock();
		try
		{
			ServiceEndpointDiscoveryOptionImpl serviceEndpointDiscoveryOptionImpl = (ServiceEndpointDiscoveryOptionImpl)params;
			_reactorErrorInfo.clear();
			_reactorServiceDiscoveryOptions.clear();
			
			_reactorServiceDiscoveryOptions.userName(serviceEndpointDiscoveryOptionImpl._userName);
			_reactorServiceDiscoveryOptions.password(serviceEndpointDiscoveryOptionImpl._password);
			_reactorServiceDiscoveryOptions.clientId(serviceEndpointDiscoveryOptionImpl._clientId);
			_reactorServiceDiscoveryOptions.clientSecret(serviceEndpointDiscoveryOptionImpl._clientSecret);
			_reactorServiceDiscoveryOptions.tokenScope(serviceEndpointDiscoveryOptionImpl._tokenScope);
			_reactorServiceDiscoveryOptions.transport(serviceEndpointDiscoveryOptionImpl._transport);
			_reactorServiceDiscoveryOptions.dataFormat(serviceEndpointDiscoveryOptionImpl._dataFormat);
			_reactorServiceDiscoveryOptions.takeExclusiveSignOnControl(serviceEndpointDiscoveryOptionImpl._takeExclusiveSignOnControl);
			
			if(serviceEndpointDiscoveryOptionImpl._proxyHostName != null && !serviceEndpointDiscoveryOptionImpl._proxyHostName.isEmpty())
			{
				_reactorServiceDiscoveryOptions.proxyHostName().data(serviceEndpointDiscoveryOptionImpl._proxyHostName);
			}
			
			if(serviceEndpointDiscoveryOptionImpl._proxyPort != null && !serviceEndpointDiscoveryOptionImpl._proxyPort.isEmpty())
			{
				_reactorServiceDiscoveryOptions.proxyPort().data(serviceEndpointDiscoveryOptionImpl._proxyPort);
			}
			
			if(serviceEndpointDiscoveryOptionImpl._proxyUserName != null && !serviceEndpointDiscoveryOptionImpl._proxyUserName.isEmpty())
			{
				_reactorServiceDiscoveryOptions.proxyUserName().data(serviceEndpointDiscoveryOptionImpl._proxyUserName);
			}
			
			if(serviceEndpointDiscoveryOptionImpl._proxyPassword != null && !serviceEndpointDiscoveryOptionImpl._proxyPassword.isEmpty())
			{
				_reactorServiceDiscoveryOptions.proxyPassword().data(serviceEndpointDiscoveryOptionImpl._proxyPassword);
			}
			
			if(serviceEndpointDiscoveryOptionImpl._proxyDomain != null && !serviceEndpointDiscoveryOptionImpl._proxyDomain.isEmpty())
			{
				_reactorServiceDiscoveryOptions.proxyDomain().data(serviceEndpointDiscoveryOptionImpl._proxyDomain);
			}
			
			if(serviceEndpointDiscoveryOptionImpl.proxyLocalHostName() != null && !serviceEndpointDiscoveryOptionImpl.proxyLocalHostName().isEmpty())
			{
				_reactorServiceDiscoveryOptions.proxyLocalHostName().data(serviceEndpointDiscoveryOptionImpl.proxyLocalHostName());
			}
			
			if(serviceEndpointDiscoveryOptionImpl._proxyKrb5ConfigFile != null && !serviceEndpointDiscoveryOptionImpl._proxyKrb5ConfigFile.isEmpty())
			{
				_reactorServiceDiscoveryOptions.proxyKRB5ConfigFile().data(serviceEndpointDiscoveryOptionImpl._proxyKrb5ConfigFile);
			}
			
			_reactorServiceDiscoveryOptions.reactorServiceEndpointEventCallback(this);
			_client = client;
			
			if(closure != null) _reactorServiceDiscoveryOptions.userSpecObj(closure);
			
			switch(serviceEndpointDiscoveryOptionImpl._transport)
			{
			case ServiceEndpointDiscoveryOption.TransportProtocol.UNKNOWN:
				break;
			case ServiceEndpointDiscoveryOption.TransportProtocol.TCP:
				_reactorServiceDiscoveryOptions.transport(ReactorDiscoveryTransportProtocol.RD_TP_TCP);
				break;
			case ServiceEndpointDiscoveryOption.TransportProtocol.WEB_SOCKET:
				_reactorServiceDiscoveryOptions.transport(ReactorDiscoveryTransportProtocol.RD_TP_WEBSOCKET);
				break;
			default:
				strBuilder().append("Invalid transport protocol ").append(serviceEndpointDiscoveryOptionImpl._transport)
					.append(" specified in ServiceEndpointDiscoveryOption.transport()");
				
				throw ommIUExcept().message(_strBuilder.toString(), OmmInvalidUsageException.ErrorCode.INVALID_ARGUMENT);
			}
			
			switch(serviceEndpointDiscoveryOptionImpl._dataFormat)
			{
			case ServiceEndpointDiscoveryOption.DataformatProtocol.UNKNOWN:
				break;
			case ServiceEndpointDiscoveryOption.DataformatProtocol.RWF:
				_reactorServiceDiscoveryOptions.dataFormat(ReactorDiscoveryDataFormatProtocol.RD_DP_RWF);
				break;
			case ServiceEndpointDiscoveryOption.DataformatProtocol.JSON2:
				_reactorServiceDiscoveryOptions.transport(ReactorDiscoveryDataFormatProtocol.RD_DP_JSON2);
				break;
			default:
				strBuilder().append("Invalid dataformat protocol ").append(serviceEndpointDiscoveryOptionImpl._dataFormat)
					.append(" specified in ServiceEndpointDiscoveryOption.dataFormat()");
				
				throw ommIUExcept().message(_strBuilder.toString(), OmmInvalidUsageException.ErrorCode.INVALID_ARGUMENT);
			}
			
			if (_reactor.queryServiceDiscovery(_reactorServiceDiscoveryOptions, _reactorErrorInfo) != ReactorReturnCodes.SUCCESS)
			{
				strBuilder().append("Failed to query service discovery (Reactor.queryServiceDiscovery).")
				.append("' Error Id='").append(_reactorErrorInfo.error().errorId()).append("' Internal sysError='")
				.append(_reactorErrorInfo.error().sysError()).append("' Error Location='")
				.append(_reactorErrorInfo.location()).append("' Error Text='")
				.append(_reactorErrorInfo.error().text()).append("'. ");
				
				throw ommIUExcept().message(_strBuilder.toString(), _reactorErrorInfo.code());
			}
		}
		finally
		{
			_userLock.unlock();
		}
	}

	@Override
	public void uninitialize()
	{	
		_userLock.lock();
		try
		{
			if(!_reactor.isShutdown())
			{
				_reactorErrorInfo.clear();
				_reactor.shutdown(_reactorErrorInfo);
			}
		}
		finally
		{
			_userLock.unlock();
		}
	}
	
	OmmInvalidUsageExceptionImpl ommIUExcept()
	{
		if (_ommIUExcept == null)
			_ommIUExcept = new OmmInvalidUsageExceptionImpl();

		return _ommIUExcept;
	}
	
	StringBuilder strBuilder()
	{
		if(_strBuilder == null)
		{
			_strBuilder = new StringBuilder(255);
		}
		else
		{
			_strBuilder.setLength(0);
		}
		
		return _strBuilder;
	}

	@Override
	public int reactorServiceEndpointEventCallback(ReactorServiceEndpointEvent event)
	{
		ReactorServiceEndpointInfo reactorServiceEndpointInfo;
		ServiceEndpointDiscoveryInfoImpl serviceEndpointDiscoveryInfoImpl;
		
		_ServiceEndpointDiscoveryEventImpl._closure = event.userSpecObject();
		_ServiceEndpointDiscoveryEventImpl._serviceEndpointDiscovery = this;
		if(event.errorInfo().code() == ReactorReturnCodes.SUCCESS)
		{
			_serviceEndpointDiscoveryRespImpl._serviceEndpointInfoList.clear();
			for(int index = 0 ; index < event.serviceEndpointInfo().size() ; index++ )
			{
				reactorServiceEndpointInfo = event.serviceEndpointInfo().get(index);
				serviceEndpointDiscoveryInfoImpl = new ServiceEndpointDiscoveryInfoImpl();
				
				for(int valueIndex = 0; valueIndex < reactorServiceEndpointInfo.dataFormatList().size(); valueIndex++)
				{
					serviceEndpointDiscoveryInfoImpl._dataFormatList.add(reactorServiceEndpointInfo.dataFormatList().get(valueIndex).toString());
				}
				
				for(int valueIndex = 0; valueIndex < reactorServiceEndpointInfo.locationList().size(); valueIndex++)
				{
					serviceEndpointDiscoveryInfoImpl._locationList.add(reactorServiceEndpointInfo.locationList().get(valueIndex).toString());
				}
				
				serviceEndpointDiscoveryInfoImpl._endpoint = reactorServiceEndpointInfo.endPoint();
				serviceEndpointDiscoveryInfoImpl._port = reactorServiceEndpointInfo.port();
				serviceEndpointDiscoveryInfoImpl._provider = reactorServiceEndpointInfo.provider();
				serviceEndpointDiscoveryInfoImpl._transport = reactorServiceEndpointInfo.transport();
				_serviceEndpointDiscoveryRespImpl._serviceEndpointInfoList.add(serviceEndpointDiscoveryInfoImpl);
			}
			
			_client.onSuccess(_serviceEndpointDiscoveryRespImpl, _ServiceEndpointDiscoveryEventImpl);
		}
		else
		{
			_client.onError(event.errorInfo().error().text(), _ServiceEndpointDiscoveryEventImpl);
		}
		
		return ReactorCallbackReturnCodes.SUCCESS;
	}
}
